source/eventcore/drivers/posix/processes.d

```diff
@@ -11,426 +11,443 @@ import std.variant : visit;
+	}
+
+	final override ProcessID adopt(int system_pid)
+	{
+		auto pid = cast(ProcessID)system_pid;
+
+		ProcessInfo info;
+		info.exited = false;
+		info.refCount = 1;
+		info.driver = this;
+		StaticProcesses.add(pid, info);
+
+		return pid;
+	}
+
+	final override Process spawn(
+		string[] args,
+		ProcessStdinFile stdin,
+		ProcessStdoutFile stdout,
+		ProcessStderrFile stderr,
+		const string[string] env,
+		ProcessConfig config,
+		string working_dir)
+	@trusted {
+		// Use std.process to spawn processes
+		import std.process : pipe, Pid, spawnProcess;
+		import std.stdio : File;
+		static import std.stdio;
+
+		static File fdToFile(int fd, scope const(char)[] mode)
+		{
+			try {
+				File f;
+				f.fdopen(fd, mode);
+				return f;
+			} catch (Exception e) {
+				assert(0);
+			}
+		}
+
+		try {
+			Process process;
+			File stdinFile, stdoutFile, stderrFile;
+
+			stdinFile = stdin.visit!(
+				(int handle) => fdToFile(handle, "r"),
+				(ProcessRedirect redirect) {
+					final switch (redirect) {
+					case ProcessRedirect.inherit: return std.stdio.stdin;
+					case ProcessRedirect.none: return File.init;
+					case ProcessRedirect.pipe:
+						auto p = pipe();
+						process.stdin = m_driver.pipes.adopt(dup(p.writeEnd.fileno));
+						return p.readEnd;
+					}
+				});
+
+			stdoutFile = stdout.visit!(
+				(int handle) => fdToFile(handle, "w"),
+				(ProcessRedirect redirect) {
+					final switch (redirect) {
+					case ProcessRedirect.inherit: return std.stdio.stdout;
+					case ProcessRedirect.none: return File.init;
+					case ProcessRedirect.pipe:
+						auto p = pipe();
+						process.stdout = m_driver.pipes.adopt(dup(p.readEnd.fileno));
+						return p.writeEnd;
+					}
+				},
+				(_) => File.init);
+
+			stderrFile = stderr.visit!(
+				(int handle) => fdToFile(handle, "w"),
+				(ProcessRedirect redirect) {
+					final switch (redirect) {
+					case ProcessRedirect.inherit: return std.stdio.stderr;
+					case ProcessRedirect.none: return File.init;
+					case ProcessRedirect.pipe:
+						auto p = pipe();
+						process.stderr = m_driver.pipes.adopt(dup(p.readEnd.fileno));
+						return p.writeEnd;
+					}
+				},
+				(_) => File.init);
+
+			const redirectStdout = stdout.convertsTo!ProcessStdoutRedirect;
+			const redirectStderr = stderr.convertsTo!ProcessStderrRedirect;
+
+			if (redirectStdout) {
+				assert(!redirectStderr, "Can't redirect both stdout and stderr");
+
+				stdoutFile = stderrFile;
+			} else if (redirectStderr) {
+				stderrFile = stdoutFile;
+			}
+
+			Pid stdPid = spawnProcess(
+				args,
+				stdinFile,
+				stdoutFile,
+				stderrFile,
+				env,
+				cast(std.process.Config)config,
+				working_dir);
+			process.pid = adopt(stdPid.osHandle);
+			stdPid.destroy();
+
+			return process;
+		} catch (Exception e) {
+			return Process.init;
+		}
+	}
+
+	final override void kill(ProcessID pid, int signal)
+	@trusted {
+		import core.sys.posix.signal : pkill = kill;
+
+		pkill(cast(int)pid, signal);
```

Comment:
Is it better to check `cast(int)pid > 0` ?
[kill(2)'s man page](http://man7.org/linux/man-pages/man2/kill.2.html) says:

> If pid equals 0, then sig is sent to every process in the process
> group of the calling process.

> If pid equals -1, then sig is sent to every process for which the
> calling process has permission to send signals, except for process 1
> (init), but see below.

> If pid is less than -1, then sig is sent to every process in the
> process group whose ID is -pid.

Reply:
Makes sense!